sled-agent/src/nexus.rs

```diff
@@ -202,3 +167,816 @@ impl ConvertInto<nexus_client::types::DatasetKind>
+
+            // Does CRDB actually contain an existing record for this sled?
+            match known_info {
+                NexusKnownInfo::NotFound => {
+                    // Nothing to do. We must send the request as is.
+                }
+                NexusKnownInfo::Found(known) => {
+                    // We don't need to send a request if the info is identical to what
+                    // nexus knows
+                    if info == *known {
+                        self.pending_notification = false;
+                        self.cancelled_pending_notifications += 1;
+                        return;
+                    }
+
+                    // Bump the generation of the value that nexus knows, so
+                    // that the update takes precedence.
+                    info.generation = known.generation.next();
+                }
+            }
+
+            let outstanding_req_ready = self.outstanding_req_ready.clone();
+            self.total_put_requests_started += 1;
+            self.outstanding_request = Some(tokio::spawn(async move {
+                let res =
+                    client.sled_agent_put(&sled_id, &info).await.map(|_| info);
+                outstanding_req_ready.notify_one();
+                (NexusOp::Put, res)
+            }));
+        } else {
+            let outstanding_req_ready = self.outstanding_req_ready.clone();
+            self.total_get_requests_started += 1;
+            self.outstanding_request = Some(tokio::spawn(async move {
+                let res = client
+                    .sled_agent_get(&sled_id)
+                    .await
+                    .map(|info| info.into_inner());
+                outstanding_req_ready.notify_one();
+                (NexusOp::Get, res)
+            }));
+        }
+    }
+
+    /// Handle a reply from nexus by extracting the value from the `JoinHandle` of
+    /// the last outstanding request.
+    async fn handle_nexus_reply(&mut self) {
+        let (op, res) = match self
+            .outstanding_request
+            .take()
+            .expect("missing JoinHandle")
+            .await
+        {
+            Ok(res) => res,
+            Err(e) => {
+                error!(self.log, "Nexus request task exited prematurely: {e}");
+                return;
+            }
+        };
+        match (op, res) {
+            (NexusOp::Get, Ok(info)) => {
+                info!(
+                    self.log,
+                    "Retrieved SledAgentInfo from Nexus: {:?}", info
+                );
+                assert!(self.nexus_known_info.is_none());
+                self.nexus_known_info = Some(NexusKnownInfo::Found(info));
+                self.total_get_requests_completed += 1;
+            }
+            (NexusOp::Put, Ok(info)) => {
+                // Unwrap Safety: we must have a known value in order to have
+                // submitted a PUT request in the first place.
```

Comment:
Thanks for spelling this out--love to see `unwrap()`s annotated this way.

Reply:
Thanks! I definitely stole this great idea from @jgallagher